operators/tokenizer/bpe_kernels.cc

```diff
@@ -0,0 +1,359 @@
+    res.push_back(bos_token_id_);
+  }
+  if (ModelName() == BpeModelConf::kModel_CLIP){
+    // Convert to lowercase
+    std::transform(input.begin(), input.end(), input.begin(), [](char32_t c) { return static_cast<char32_t>(ToLower(c)); });
+  }
+
+  // Parse input
+  auto special_token_split_res = bbpe_tokenizer_->SplitBySpecialTokens(input);
+  TokenWithRegularExp regcmp;
+
+  for (auto& seg_id : special_token_split_res) {
+    if (static_cast<int64_t>(res.size()) >= max_length) break;
+
+    if (seg_id.second != -1) {
+      res.push_back(seg_id.second);
+      continue;
+    }
+
+    auto cur_input = std::move(seg_id.first);
+    // Note: keep ptr to make sure the string_view is valid in the following process
+    const char32_t* ptr = cur_input.c_str();
+    regcmp.Set(ptr);
+
+    size_t offset = 0;
+    OffsetMappingType offset_mapping;
+
+    if (compute_offset_mapping) {
+      if (ModelName() != BpeModelConf::kModel_GPT2) {
+        // Add offset mapping for BOS token
+        offset_mapping.push_back(std::make_pair(0, 0));
+      }
+    }
+
+    while (static_cast<int64_t>(res.size()) < max_length) {
+      auto [b, tok] = regcmp.GetNextToken();
+      if (!b) break;
+
+      std::string utf8_token = std::string(ustring(tok));
+
+      size_t space_dif = 0;
+      if (compute_offset_mapping) {
+        // Handle special case for offset mapping
+        if (utf8_token.at(0) == ' ') {
+          offset++;
+          space_dif = -1;  // account for spaces used in offset map algorithm in bpe(byte_list_)
+        }
+      }
+
+      // Get byte encodings prior to performing BPE
+      byte_list.clear();
+
+      if (ModelName() == BpeModelConf::kModel_CLIP) {
+        // Whitespace clean
+        utf8_token.erase(std::remove(utf8_token.begin(), utf8_token.end(), ' '), utf8_token.end());
+
+        for (int i = 0; i < utf8_token.length(); i++) {
+          if (i == utf8_token.length() - 1) {
+            std::string boundary(1, utf8_token[i]);
+            byte_list.push_back(std::make_pair(bbpe_tokenizer_->GetTokenId(boundary + "</w>"), 1));
+          } else {
+            byte_list.push_back(std::make_pair(bbpe_tokenizer_->ByteEncoder()[static_cast<unsigned char>(utf8_token[i])], 1));
+          }
+        }
+      } else {
+        for (char& cp : utf8_token) {
+          byte_list.push_back(std::make_pair(bbpe_tokenizer_->ByteEncoder()[static_cast<unsigned char>(cp)], 1));
+        }
+      }
+
+      // Perform BPE
+      bbpe_tokenizer_->bpe(byte_list);
+
+      // Add output to result
+      for (auto p : byte_list) {
+        if (static_cast<int64_t>(res.size()) >= max_length) {
+          break;
+        }
+
+        res.push_back(p.first);
+
+        if (compute_offset_mapping) {
+          if (ModelName() == BpeModelConf::kModel_CLIP) {
+            offset_mapping.emplace_back(std::make_pair(offset, ort_extensions::narrow<size_t>(offset + p.second)));
+            offset += p.second;
+          } else {
```

Comment:
should we add a check for ModelName() == BpeModelConf::kModel_Roberta here because the offset map alg might not apply to others (in case we add more BPE-based tokenizers in the future)?

or we could just add a comment to say that it is for Roberta - "If adding new BPE tokenizers, check for a different offset mapping algorithm".

Reply:
I assume this behavior applies all models which don't clean up white-space like CLIP, not only Roberta.